### PR TITLE
Using absolute path and installable package classes

### PR DESCRIPTION
**Pull request by akshita31**
At a number of places in the code, we require that the paths that are associated with the packages must be absolute. However the package interface imposed no such restrictions which made the download and install pipeline error prone. 
Hence added an `InstallablePackage` class that requires that the `installPath`, `installTestPath` and the `binaries` must have absolute and not relative paths.

### CODE REVIEW

File: src/packageManager/InstallablePackage.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+function getAbsoluteBinaries(pkg: Package, extensionPath: string): AbsolutePath[] { 
+    let basePath = getAbsoluteInstallPath(pkg, extensionPath).value;
+    if (pkg.binaries) { 
+        return pkg.binaries.map(value => AbsolutePath.getAbsolutePath(basePath, value)); 
+    } 
+ 
+    return null; 
+} 
+ 
+function getAbsoluteInstallPath(pkg: Package, extensionPath: string): AbsolutePath {
```
Comment (rchande) [inline src/packageManager/InstallablePackage.ts]: extensionPath is kind of a confusing name. I think this is the base directory of the extension (or something?) but it winds up being the `pathToPrepend` despite being called extension?

File: src/packageManager/Package.ts [hunk 2]
```diff
@@ -3,15 +3,17 @@
...
     fallbackUrl?: string;
-    installPath?: string;
     platforms: string[];
     architectures: string[];
-    binaries: string[];
     platformId?: string;
-    // Path to use to test if the package has already been installed
+}
+
+export interface Package extends IPackage{
```
Comment (rchande) [inline src/packageManager/Package.ts]: `RelativePathPackage`?

File: src/packageManager/InstallablePackage.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { IPackage, Package } from "./Package";
+import { AbsolutePath } from "./AbsolutePath";
+
+export class InstallablePackage implements IPackage{
```
Comment (rchande) [inline src/packageManager/InstallablePackage.ts]: `AbsolutePathPackage`?

File: src/packageManager/AbsolutePath.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import { isAbsolute, resolve } from "path";
+
+export class AbsolutePath{
+    constructor(public value: string) {
+        if (!isAbsolute(value)) {
+            throw new Error("The path must be absolute");
+        }
+    }
+
+    public static getAbsolutePath(pathToPrepend: string, path: string): AbsolutePath {
```
Comment (rchande) [inline src/packageManager/AbsolutePath.ts]: thoughts on making the signature here just look like the signature of path.resolve?

File: src/packageManager/Package.ts (hunk 2 above)
Comment (akshita31) [reply src/packageManager/Package.ts]: I am not sure we should do that. The paths here can be absolute or relative there is no specific restriction that we are imposing right(like isRelative(path) or something). Thoughts ?

File: src/packageManager/AbsolutePath.ts (hunk 4 above)
Comment (akshita31) [reply src/packageManager/AbsolutePath.ts]: Done

File: src/packageManager/InstallablePackage.ts (hunk 3 above)
Comment (akshita31) [reply src/packageManager/InstallablePackage.ts]: Done

**Pull request merged**